### ISSUE TITLE
some waypoints promote bad accessibility practices

### ISSUE THREAD

**Issue by zersiax** [FreeCodeCamp/FreeCodeCamp#4656]
I just came across the following line of code in the second bootstrap waypoint:
 `<a href="#"><img class="smaller-image thick-green-border" src="https://bit.ly/fcc-relaxing-cat"></a> `
I cannot in good  conscience  ignore this. A screenreader will read the resulting output in the following, rather helpful manner:
"link graphic
I appreciate the waypoint about the alt-tag being included in the FCC curriculum, please let there be no doubt about that. But I do think if FCC wants to teach proper HTML/CSS, a grave mistake like this should not be allowed to exist and the advice in the waypoint about the alt-tag should be taken for FCC's waypoints, ziplines, bonfires and oher exercises as well.
I also tripped over the lack of the 'for'-attribute in the label tag for forms. I suppose I can forgive that one for now since that attribute goes into some rather advanced semantic discussions and is exclusively for semantic correctness and accesibility, but I do hope this topic will be discussed at a later part in the curriculum. If not I'll have to write up a few waypoints meself, I reckon ;)
For completeness, the URL where I came across the offending line of code is below:
http://www.freecodecamp.com/challenges/waypoint-make-images-mobile-responsive

**Comment by ltegman** [comment]
I'm completely in agreement with this. The lesson in question comes after students have completed the lesson on alt tags and we should do everything we can to make it so that when students think about creating an `img` tag they see `alt` tags as integral to the process.

**Comment by zersiax** [comment]
I'v already considered writing up articles and wapoints about
accessibility in web applications for FCC (use of proper HTML5
semantics, ARIA as a fallback option etc). This community has given me
a lot and I'm more than willing to give back. We do first need to get
these basics hammered in well and deep though. :)

2015-11-22 23:18 GMT+01:00, Logan Tegman <notifications@github.com>:

**Comment by SaintPeter** [comment]
I agree, this is an important part of modern web design.  It sounds like at very least we need some new waypoints to cover it and then to bake it into subsequent lessons.

@zersiax If this is something you're passionate about, maybe you can help?

**Comment by zersiax** [comment]
Naturally :) I'll clone the repo later today and see what can be done. I think the most important part right now is to make sure lessons that already exist have these practices included. I understand not everyone's up to snuff on the latest tips and tricks in this field, but adding alt attributes has been around for a while now :) FCC has a lesson on it and he way it is implemented now, i.e. mention it but not using it ourself, has the added problem of labeling it as something of an afterthought or extra thing to do. That's a dangerous mindset because this is indeed how many devs think about accessibility, something which is coming back around to bite them now more and more public sector websites for example need to comply with the WCAG AA specs.

**Comment by zersiax** [comment]
Just had a look at the code for challenges, I shouldn't have a lot of trouble modifying it. What branch do I work in, still staging? Or do I make a separate branch for this?

**Comment by SaintPeter** [comment]
Our [Contribution Guidelines](https://github.com/FreeCodeCamp/FreeCodeCamp/blob/staging/CONTRIBUTING.md) should answer your questions.  Please let me know if you have any further questions.

**Comment by zersiax** [comment]
Can't seem to find how to assign this issue to me. Can't seem to do that from either the github issue page or the waffleboard, guess you have to be the repository owner to do that?

**Comment by SaintPeter** [comment]
Unless you have write access to the main repo you cannot be assigned.  That's ok, though, it's unlikely that someone will snatch this out from under you.

**Comment by bigkatspence** [comment]
As a current camper I have seen a couple of the waypoints where the end
code has errors that codepen

**Comment by zersiax** [comment]
It's been a while since I last updated this ticket, simply because I
just don't have the time I expected to have to fix this.
As discussed in the Curriculum development gitter channel, I am very
willing to still write the wiki articles on web accessibility.
However, I've done some thinking.
If I were to fix these issues all by myself and then leave you guys to
it, nobody will have learned a thing.  If the challenge map will be
overhauled in the future, which is very possible, and others will
rewrite these HTML challenges for whatever reason, we may be back to
square one sooner than we might like.
Therefore I propose someone of the core team picks this one up. If you
want to do this web accessibility business right, it has to be
interwoven through the development process of the entire system, not
just these few waypoints. If however you don't really care, or don't
think you have the time, or don't think it's a priority like most
initiatives out there that is of course your prerogative. In that case
however I don't really feel like being the only one who patches the
holes that will eventually show up. I'd like to start a dialogue on
this. What say you?

**Comment by ltegman** [comment]
I don't think framing your contributing better accessibility practices as "patching up" other people's ill-considered work is the right way to think about this. We're an open source community full of people with different strengths and interests. If one of us contributes lessons that are effective teaching tools and you contribute better accessibility practices after the fact that's the beauty of open-source collaboration. We can all come together and contribute our strengths to make an end product that is better than any one of us could make alone.

Are there times your work will get accidentally removed and a lesson will take a step backward in some regard? Yep. There's a lot of moving parts and stuff like that is bound to happen from time to time. Such is the nature of a big project like this.
Should everyone try to be better about accessibility? Sure. But I don't think whether or not that happens should deter you from contributing what you can.

We all contribute what we're willing and able in the time we have. If time and a will to contribute are something you have I encourage you to do so. If not, no one will begrudge you that :smile:

**Comment by zersiax** [comment]
What I aim to contribute in that regard is easy to follow guidelines
and tips in the earlier mentioned articles, so this accessibility
thing will become less of a hassle and more of a common sense kind of
thing.
I don't have an issue with patching up other people's work after the
fact so much as these blatant accessibility flaws existing just
because one person, me in this case, doesn't have the time to fix
them. Especially things like missing alt tags, which are actually even
a breach of the HTML5 specification and not just an accessibility
problem, should be easy to fix for anyone who would like to contribute
to FCC and it's quality and accessibility. I'd love to help make FCC
the best it can be, but right now what I am doing feels like a bit of
an afterthought while I think we should work on getting these things
integrated in the core processes so this patching up after the fact
isn't necessary (as much) anymore. Of course, accessibility
regressions can always occur, but getting a number of a11y checks
integrated in the core processes will prevent a lot of them. For
example, pull requests need to be submitted to add new waypoints.
Someone has to accept those pull requests. The way I see it, at this
point the person who merges the pull request should know what to look
for and that includes things like quality of submission, spelling
mistakes, educational value, conformance with the relevant
specification (HTML5 spec, JSLint etc) and some very simple
accessibility checks like missing alt-tags. The reason I think this is
so important for a project like FreeCodeCamp is because it aims to
teach people who (often) will want to find a job with the skills they
have learnt. FCC, in my opinion, has a responsibility to make sure
what is taught is actualy semantically correct and of sufficient
quality. To teach something and not apply it in subsequent lessons
even though in the relevant lesson it's importance is emphasized does
not send out the right signals in my opinion and brings me back to
this afterthought feeling I spoke of, which is a dangerous mindset
because not adhering to these specs can have serious consequences for
the companies in question. What are your thoughts on this?

**Comment by SaintPeter** [comment]
At some point I will be doing a revamp on the entire HTML/CSS curriculum.  It sounds like I need to get up to speed on accessibility best practices so when I do so I can update them.

I get where you're coming from - you really would be playing cleanup and if we don't have an across the board commitment to integrating accessiblity into the curriculum, you'll always be one step behind.

That said . . it's already written.  What is done is done.  If we're going to have a commitment to accessibility, we're GOING to have to clean up what we've got.

Is it something as simple as a checklist?  Or is it deeper than that?

**Comment by zersiax** [comment]
It kind of depends on the complexity of the exercise, really.
When you use standard semantic HTML according to the HTML5 spec, you
are pretty much 90% of the way there already in most cases, especially
if we are just talking HTML/CSS. Things get more interesting when JS
gets involved, but I'll leave that for the articles I'm about to write
:) In the case of HTML/CSS, I'd say the points to look at are:
- When using headings, use them hierarchically. If you put an h3 below
an h2, make sure that actually makes sense. e.g. an h2 above a list of
posts and an h3 for each post. How NOT to do it is for example to make
an h2 called 'Everything you always wanted to know about cats' and
then add an h3 with "contact us", which could for example be in the
footer,  right below, unless you think that is actually part of the
main content, but in this case I think you can see that wouldn't make
any sense :)
- That brings me to my next point. When writing HTML/CSS, make sure
your site is still readable and logical when your CSS is turned off.
Screenreaders, search engines and also other assistive tech infers the
order of the HTML document from the HTML source, CSS is often stripped
off almost 100%. So if you put your footer first in the HTML and just
position it at the bottom with CSs, screenreaders will read the footer
first. Which tends not to be what you want :) The same goes for only
denoting important content with a CSS color style or a font-awesome
icon. You can do that, but make sure there's a text alternative, for
example using a span that lives off-screen like I did for the
'completed' and 'not completed' for the challenge map on FCC.
- Last gotcha is images. Images always require an alt-tag. This is not
an accessibility rule per se, its an HTML5 spec thing. There are
certainly cases though where this alt tag may be left blank. If the
image is purely decorative, contains no extra information at all and
is sooner to bother a screenreader user than help one in your opinion,
that's when you use alt="". In pretty much every other case a
description is rather strongly suggested.

Most of this is what I will be putting in my first accessibility
article, in a bit more detail. I should have that written up some time
next week. :)
Finally, please understand that I don't mind cleaning up what's
already written. I think it would be a good exercise for someone else
to do it and that would certainly make it go faster since i don't
really have the time for it now, but I'd be happy to clean it up. I'm
just rather hesitant to do that because I know FCC changes often and I
think in this case solving the problem and not fighting the symptoms
will be a far better thing to do in the long run. If there are any
more questions please do ask :)


2015-12-30 7:34 GMT+01:00, Rex Schrader <notifications@github.com>:

**Comment by erikasf** [comment]
Hi,
If this is still an issue, I would be thrilled to help out. This is an area I am passionate about and  in fact I am giving a talk  on this exact subject at the world dev conference and providing a couple of seminars through the dev network. I would be more than happy to help out, if there is a need.

**Comment by zersiax** [comment]
Hi Erika,

Yup, the issue is still open since I sadly don't have the time to sift
through the JSON and fix missing alts etc. :)
Send me am email if you have any questions :)

2015-12-30 18:18 GMT+01:00, erika <notifications@github.com>:

**Comment by RachelRVasquez** [comment]
Hi there! I was just about to leave an issue when I saw this existing one. I'd like to help edit or create new waypoints to promote better a11y practices as well. I was just going through the basics when I reached the waypoint promoting the use of hash tags in anchor links and thought it'd be a good opportunity to teach others not to go this route. Instead, to use button elements. Here's a reference:

http://www.karlgroves.com/2013/05/14/links-are-not-buttons-neither-are-divs-and-spans/

Let me know how I can help. I'm still learning about accessibility, but have been trying to share what I've been learning this far on my blog: 

http://rachievee.com/category/wordpress/accessibility-series/

Forking the project today to take a look and see where I can assist. :+1:

**Comment by buckeyekt** [comment]
Hi!

I'm just learning to code and began using FreeCodeCamp to learn some front end dev. Wanted to try my hand at some pull requests. I was a Special Educator for 10 years so accessibility is near and dear to my heart, though I know nothing about how to code for it. I am very passionate about learning how to code for accessibility from the start. Let me know if there are still waypoints that I can help clean up. I would love to contribute to this issue, and learn by doing it.

Thanks!

Katie

**Comment by zersiax** [comment]
Hi,

If nobody else pull-requested about this yet, than I think there still
will be accessibility issues especially in the HTML/CSS curriculum.
I promised to write a few articles about accessibility for the FCC
wiki, I am working on them :) I should have the first one done by
tonight, after which you can learn how to fix at least those waypoints
up. That first article will consist of what I think are the basics of
any accessibility layer in a web application.


2016-01-16 6:18 GMT+01:00, buckeyekt <notifications@github.com>:

**Comment by krisgesling** [comment]
Hi all, just been lurking a little as I think accessibility is such a critical component to add to the curriculum but don't have much time at this moment to help make it happen. 

@buckeyekt - there are a few places to get more information if you want to read up on it
[WebAIM](http://webaim.org/intro/) is an easy intro that gives you the broad picture with some of the most common accessibility problems.
[AccessIQ](http://www.accessiq.org) is an Australian based group that has good content including specific guides for devs, designers and content writers, but a lot of it is 'premium' paid content.
Then there's always the official [Web Content Accessibility Guidelines](https://www.w3.org/TR/WCAG20/) if you don't mind diving straight into the documentation.

**Comment by zersiax** [comment]
To everyone wanting to learn about accessibility basics, the first of
my articles just went live on the FCC wiki at
https://github.com/FreeCodeCamp/FreeCodeCamp/wiki/accessibility-101
Please note there are still a few minor cosmetic issues to fix, I am
working on that as we speak.

**Comment by zersiax** [comment]
Hi,

Just checking in on this issue to see if anyone has picked this up
yet? Also, are there any questions that my article doesn't quite cover
when it comes to improving the FCC HTML curriculum?

**Comment by matty22** [comment]
@zersiax I started looking at it yesterday, but ran into the roadblock of getting the challenge editing tool installed. I am not an accessibility expert, so my contribution to this issue is likely to be adding alt tags to all the images in the HTML challenges.

Once I've gotten the tool installed, it should be a quick job of adding those alt tags.

**Comment by zersiax** [comment]
That would already be a great help. I can probably do the rest later :-)

**Comment by zersiax** [comment]
Hi,

I probably have some time to work on this now, did you get anywhere
with adding the alts? :)

**Comment by matty22** [comment]
@zersiax No, I couldn't ever get the challenge editor software installed properly so I've just shelved doing any challenge related contributions for now.

**Comment by BerkeleyTrue** [comment]
This work has moved to https://github.com/FreeCodeCamp/CurriculumExpansion
